### PR TITLE
MULE-14171: SDK loses the Map DataType generics

### PR DESCRIPTION
**Pull request by estebanwasinger**

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/AbstractReturnDelegate.java [hunk 1]
```diff
@@ -105,6 +121,9 @@ protected Message toMessage(Object value, ExecutionContextAdapter operationConte
       // org.mule.runtime.api.metadata.DataType.MULE_MESSAGE_COLLECTION doesn't completely makes sense
       if (returnsListOfMessages && value instanceof Collection) {
         messageBuilder = Message.builder().collectionValue((Collection) value, Message.class);
+      } else if (isMapType && value instanceof Map) {
+        messageBuilder =
+            Message.builder().mapValue((Map) value, mapKeyType, mapValueType);
```
Comment (afelisatti) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/AbstractReturnDelegate.java]: Why the extra line?

File: core/src/main/java/org/mule/runtime/core/internal/message/InternalMessage.java [hunk 2]
```diff
@@ -239,79 +239,12 @@ static Builder builder(Message message) {
     InternalMessage build();
   }
 
-  interface CollectionBuilder extends Message.CollectionBuilder, Builder {
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/internal/message/InternalMessage.java]: Why?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/AbstractReturnDelegate.java (hunk 1 above)
Comment (estebanwasinger) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/AbstractReturnDelegate.java]: formatter /shrug

File: core/src/main/java/org/mule/runtime/core/internal/message/DefaultMessageBuilder.java [hunk 3]
```diff
@@ -122,34 +125,34 @@ public DefaultMessageBuilder(org.mule.runtime.api.message.Message message) {
...
   @Override
   public InternalMessage.Builder value(Object value) {
-    this.payload = new TypedValue(value, resolveDataType(value));
+    this.payload = new TypedValue<>(value, resolveDataType(value));
     return this;
   }
 
   @Override
-  public InternalMessage.CollectionBuilder mediaType(MediaType mediaType) {
+  public InternalMessage.MapBuilder mediaType(MediaType mediaType) {
```
Comment (alegmarra) [inline core/src/main/java/org/mule/runtime/core/internal/message/DefaultMessageBuilder.java]: why remove it from the collection but add it to the map?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/AbstractReturnDelegate.java [hunk 4]
```diff
@@ -68,6 +76,14 @@ protected AbstractReturnDelegate(ComponentModel componentModel,
                                    CursorProviderFactory cursorProviderFactory,
                                    MuleContext muleContext) {
     returnsListOfMessages = componentModel instanceof HasOutputModel && returnsListOfMessages((HasOutputModel) componentModel);
+
+    isMapType = componentModel instanceof HasOutputModel && isMap(((HasOutputModel) componentModel).getOutput().getType());
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/AbstractReturnDelegate.java]: follow the spirit of `returnsListOfMessages` and add the util method there. Also make the `instanceOf HasOutputModel` inside the methods

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessorTestCase.java [hunk 5]
```diff
@@ -229,6 +236,29 @@ public void operationReturnsPayloadValue() throws Exception {
     assertThat(message.getPayload().getValue(), is(sameInstance(value)));
   }
 
+  @Test
+  public void operationReturnsMapWithCorrectDataType() throws Exception {
+    Object payload = new HashMap<>();
+
+    Type type = new TypeToken<Map<String, String>>() {}.getType();
+    MetadataType load = new DefaultExtensionsTypeLoaderFactory().createTypeLoader().load(type);
```
Comment (alegmarra) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessorTestCase.java]: buen generated name

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessorTestCase.java [hunk 6]
```diff
@@ -229,6 +236,29 @@ public void operationReturnsPayloadValue() throws Exception {
...
+    when(operationExecutor.execute(any(ExecutionContext.class)))
+        .thenReturn(just(payload));
+
+    Message message = messageProcessor.process(event).getMessage();
+    assertThat(message, is(notNullValue()));
+
+    assertThat(message.getPayload().getValue(), is(sameInstance(payload)));
+    DataType dataType = message.getPayload().getDataType();
+    assertThat(dataType, instanceOf(MapDataType.class));
+    assertThat(((MapDataType) dataType).getKeyDataType(), like(String.class, ANY.withCharset(null)));
```
Comment (alegmarra) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessorTestCase.java]: test for different MediaTypes

File: core/src/main/java/org/mule/runtime/core/internal/message/DefaultMessageBuilder.java (hunk 3 above)
Comment (estebanwasinger) [reply core/src/main/java/org/mule/runtime/core/internal/message/DefaultMessageBuilder.java]: `badmine`

**Pull request merged**